### ISSUE TITLE
Candidate missing from /candidates/totals

### ISSUE THREAD

**Issue by jwchumley** [fecgov/openFEC#3768]
@jwchumley commented on [Wed May 15 2019](https://github.com/fecgov/openFEC/issues/3767)

@fec-issues-bot commented on [Mon May 13 2019](https://github.com/fecgov/FEC/issues/8914)

## What were you trying to do and how can we improve it?
 As A candidate for US President myself, I am disappointed and confused as to why my name James Ogle for President campaign isn't listed there anymore though I did file Form 2 and had been on this list only a few weeks ago. 

## Tell us about yourself
 Pure proportional representation (PPR) Electoral College USA. 

## Details
* URL: https://www.fec.gov/data/candidates/president/?election_year=2020&cycle=2020&election_full=true&party=GRE 
* User Agent: Amazon CloudFront

---

@jwchumley commented on [Wed May 15 2019](https://github.com/fecgov/openFEC/issues/3767#issuecomment-492662027)

WHen I add the correct candidate ID the page returns no results:

https://www.fec.gov/data/candidates/president/?q=P60004322&election_year=2020&cycle=2020&election_full=true&party=GRE

It looks like this is the api call when you search for P60004322 on the candidates/president page:

https://api.open.fec.gov/v1/candidates?api_key=5yyI90SU3Xb73TVlv4wrEhQxYcCwMWCywQiGdYbJ&candidate_id=P60004322

The API call returns a result but the page doesn't

**Comment by lbeaufort** [comment]
Thanks @jwchumley! Looks like the data is missing from candidates/totals: https://api.open.fec.gov/v1/candidates/totals/?api_key=DEMO_KEY&sort_hide_null=false&sort_nulls_last=true&q=P60004322&election_year=2020&cycle=2020&election_full=true&sort=-receipts&per_page=30&page=1&office=P

I'll go ahead and move this to the openFEC repo

**Comment by hcaofec** [comment]
Summary of how this candidate looks like in website:  **OGLE, JAMES ORLANDO III,  P60004322**

1). From "All candidates" --- Candidate is shown
https://www.fec.gov/data/candidates/?election_year=2020&office=P&party=GRE

<img width="1134" alt="Screen Shot 2019-05-28 at 12 32 30 PM" src="https://user-images.githubusercontent.com/24396726/58495450-81af8100-8145-11e9-963f-90a80b4e24c6.png">

After removing filter "Has raised more than $0"

<img width="1411" alt="Screen Shot 2019-05-28 at 12 36 02 PM" src="https://user-images.githubusercontent.com/24396726/58495455-85db9e80-8145-11e9-89af-816949d46ec5.png">

**Comment by lbeaufort** [comment]
@hcaofec the user was looking at the presidential candidates datatable, not the candidates datatable.

User was looking at https://www.fec.gov/data/candidates/president/?election_year=2020&cycle=2020&election_full=true&party=GRE

**Comment by hcaofec** [comment]
2).  From "Presidential candidates" ---- Candidate is not shown
https://www.fec.gov/data/candidates/president/?election_year=2020&cycle=2020&election_full=true&party=GRE

<img width="1186" alt="Screen Shot 2019-05-28 at 12 44 45 PM" src="https://user-images.githubusercontent.com/24396726/58495837-7d379800-8146-11e9-9257-a37744297c25.png">


<img width="1401" alt="Screen Shot 2019-05-28 at 12 56 51 PM" src="https://user-images.githubusercontent.com/24396726/58496562-20d57800-8148-11e9-92a3-a9fe85561c7f.png">

**Comment by hcaofec** [comment]
3). From "Election page" -----  Candidate is not shown
https://www.fec.gov/data/elections/president/2020/

**Comment by hcaofec** [comment]
@lbeaufort  I was comparing the candidate datatable,  presidential candidates datatable and presidential election datatable for this candidate, and try to see how an overall picture of how this candidate look like in our website

**Comment by lbeaufort** [comment]
@hcaofec sorry about that!

**Comment by hcaofec** [comment]
This candidate only filed F2 to FEC, no committee (F1) information. Thus there is no data in cand_cmte_linkage, and no financial reports.   Presidential candidates datatable and election datatable rely on financial data, so this candidate is not shown on these tow pages, but shows on 'all candidates datatable' page.   'all candidates datatable' page relies on cand_valid_fec_yr table.

@lbeaufort @jwchumley

**Comment by dorothyyeager** [comment]
Tagging @PaulClark2 to decide next steps if any with this issue.

**Comment by PaulClark2** [comment]
@dorothyyeager  from Slack #product_management

```
Laura Beaufort [12:07 PM]
re: https://fecgov.slack.com/archives/C3W9XBBJL/p1559221667004400

@hcao - I asked @paul about showing $0 in /candidates/totals/ for candidates without committees and he said that was fine - that way the list of folks running in a given office will match in the “All candidates” and “Presidential/House/Senate Candidates” datatables.

cc: @Jonella
Jonella Culmer
@here What’s the difference between the list of presidential candidates on the All candidates datatable, currently showing 740, and the list of candidates on the Presidential candidates datatable, currently showing 438? Why the difference?
Posted in #devToday at 9:07 AMView message
```

**Comment by dorothyyeager** [comment]
Closing this issue as we've resolved what we'll do about candidates with no committees (see Slack message above).

**Comment by lbeaufort** [comment]
@dorothyyeager I think this is the implementation issue, it's not done yet.

**Comment by dorothyyeager** [comment]
Thanks @lbeaufort  - there was some confusion on my end for sure!

**Comment by lbeaufort** [comment]
@dorothyyeager no worries! Thanks for staying on top of the tickets! 😃

**Comment by hcaofec** [comment]
Next step: refactor ofec_candidate_totals_mv

**Comment by jason-upchurch** [comment]
@hcaofec with cc to @lbeaufort,would it be ok with you if I add to the `pytest` suite in `test_candidates.py` further integration tests that cover this scenario once your PR is ready?

**Comment by jason-upchurch** [comment]
@hcaofec @lbeaufort I put in issue #3806 to add test coverage so I would not forget. I put it Sprint 9.2 and as "bug" but can change as necessary.

**Comment by fecjjeng** [comment]
ofec_candidate_totals_mv has two major components, one from financial side, one from the candidate site.
There are two major types of candidates that caused the mismatch of these two worlds. 
- candidates has committees but did not file financial reports
- candidates has no committees at all

We had noticed the candidate_election_year logic has problem.
In previous ticket #3700, we had fixed the financial side.  
In another ticket #3736, the candidate side had been fixed.
This ticket we combined the two sides together with some minor updates
we noticed from previous tickets.

**Comment by fecjjeng** [comment]
While working on this ticket, another ticket #3709 Candidate totals incorrect for one candidate ("Puerto Rico's resident commissioner has a 4-year election cycle. Even though the commissioner serves in the house.") was opened.  Since it involved the same sets of MVs, it is investigated together. A new function candidate_election_duration(p_candidate text) is created Puerto Rico's resident commissioner has a 4-year election cycle, even though the commissioner serves in the house.  The ofec_cand_cmte_linkage_mv and ofec_candidate_history_mv are now calling this new function to better handle the calculation of candidate_election_year; including the situation for PR house candidates.

**Comment by fecjjeng** [comment]
It is also noticed that the ofec_candidate_details_mv (which derived from ofec_candidate_history_mv) will need to be updated to better fit the new candidate_election_year calculation ticket #3736.  We have two possible way to handle it, neither way will affect the candidate list though since the API use the election_years array to handle the search.  The two proposed way to handle is illustrated by Clinton's example.  She run for president on year 2008, and year 2016.  After 2016, she still have financial activity so she still has two_year_period after 2016.  
ofec_candidate_detail_mv has one row per candidate.

ofec_candidate_detail_mv (current): The candidate's latest row of the fec_election_cycle (candidate_election_year might be empty)
ofec_candidate_detail_mv_tmp_jj(jj): The candidate's latest row of the candidate_election_cycle
ofec_candidate_detail_mv_tmp_jj(jj): The candidate's latest row of the fec_election_cycle, with the candidate_election_year field be the latest candidate_election_year this candidate had run.

select 'current', * from ofec_candidate_detail_mv where candidate_id = 'P00003392'
union
select 'jj', * from ofec_candidate_detail_mv_tmp_jj where candidate_id = 'P00003392'
union
select 'jj2', * from ofec_candidate_detail_mv_tmp_jj2 where candidate_id = 'P00003392'
order by 1

<img width="1350" alt="Screen Shot 2019-06-05 at 6 51 12 PM" src="https://user-images.githubusercontent.com/24943130/58995647-042de580-87c3-11e9-81ec-c426a5da17e7.png">

**Comment by fecjjeng** [comment]
According to research by @hcaofec 
these classes using this mv
webservices/resources/
candidates.py
CandidateList
CandidateSearch
CandidateView
----------------------------------------
Here are the endpoints that are using these resources:
api.add_resource(
   candidates.CandidateView,
   '/candidate/<string:candidate_id>/',
   '/committee/<string:committee_id>/candidates/',
)

api.add_resource(candidates.CandidateList, '/candidates/')
api.add_resource(candidates.CandidateSearch, '/candidates/search/')
There are four endpoints using ofec_candidate_details_mv ( directly or indirectly )

**Comment by fecjjeng** [comment]
@lbeaufort @PaulClark2 @jwchumley Please let us know which version of ofec_candidate_detail_mv you think would better fit the new website's needs.  I have a WIP PR #3808 that include version ofec_candidate_detail_mv_tmp_jj2 (all three versions are in dev database).
